#### Q1. What is MongoDB? Explain non-relational databases in short. In which scenarios it is preferred to use MongoDB over SQL databases?

MongoDB is a popular NoSQL (non-relational) database management system. NoSQL databases are designed to handle and store data in ways that differ from traditional relational databases, like SQL databases. Here's a brief explanation of non-relational databases and when MongoDB is preferred over SQL databases:

1. **Non-Relational Databases (NoSQL):**
   Non-relational databases, often referred to as NoSQL databases, are a category of database systems that are not based on the traditional relational database model. Unlike SQL databases, which use structured tables and a fixed schema, NoSQL databases are more flexible and can store data in various formats, such as key-value pairs, documents, graphs, or wide-column stores. They are suitable for handling large volumes of unstructured or semi-structured data and can scale horizontally to accommodate growing datasets.

2. **When to Prefer MongoDB over SQL Databases:**
   MongoDB is preferred over SQL databases in specific scenarios:

   a. **Flexible Schema:** MongoDB allows for a flexible schema, which means that each document in a collection can have a different structure. This flexibility is beneficial when dealing with rapidly evolving data or semi-structured data where the schema is not well-defined.

   b. **Scalability:** MongoDB is designed for horizontal scalability, making it suitable for applications that need to handle high volumes of data and traffic. It can distribute data across multiple servers or clusters, providing excellent scalability and performance.

   c. **Complex Data Structures:** MongoDB excels at handling complex data structures, such as nested arrays and embedded documents. This makes it a good choice for applications with data that doesn't fit neatly into the rows and columns of a traditional SQL table.

   d. **Geospatial Data:** MongoDB has built-in support for geospatial data and queries, making it a strong choice for applications that involve location-based services and mapping.

   e. **Rapid Development:** MongoDB is often favored for rapid development and prototyping because of its ease of use and flexible document-oriented data model.

   f. **Big Data and Real-Time Analytics:** For applications that require real-time data analysis and processing, MongoDB can be a suitable choice due to its ability to handle high write and read loads.

   g. **No Fixed Schema:** If your project doesn't have a well-defined schema or requires frequent schema changes, MongoDB's schema-less approach can simplify development and maintenance.

However, it's essential to note that the choice between MongoDB and SQL databases depends on the specific requirements and characteristics of your project. SQL databases are still well-suited for applications with a well-defined schema and complex relationships between data, where data consistency and ACID transactions are critical. MongoDB's strengths shine in scenarios where flexibility, scalability, and handling of unstructured or semi-structured data are paramount.

#### Q2. State and Explain the features of MongoDB.

MongoDB is a popular NoSQL database management system known for its flexibility, scalability, and ease of use. Here are some of its key features along with explanations:

1. **Flexible Schema:**
   - **Explanation:** MongoDB uses a schema-less or schema-flexible approach, allowing documents in a collection to have different structures. This flexibility makes it suitable for storing data with evolving or variable schemas, which is common in modern applications.

2. **Document-Oriented:**
   - **Explanation:** MongoDB stores data in BSON (Binary JSON) documents, which are self-contained and can contain nested arrays and subdocuments. This document-oriented data model is more intuitive for developers working with JSON-like data and simplifies data storage and retrieval.

3. **Highly Scalable:**
   - **Explanation:** MongoDB is designed for horizontal scalability, enabling it to handle large datasets and high levels of traffic. It supports sharding, which allows data distribution across multiple servers or clusters, ensuring efficient scaling.

4. **Automatic Failover and Replication:**
   - **Explanation:** MongoDB provides built-in support for replication and automatic failover. This ensures data redundancy and high availability by maintaining multiple copies (replicas) of data across different servers. If one server fails, MongoDB automatically switches to a healthy replica.

5. **Rich Query Language:**
   - **Explanation:** MongoDB offers a powerful and expressive query language that supports complex queries, indexing, and aggregation operations. Developers can perform queries on documents using a wide range of operators, including geospatial and text search queries.

6. **Indexes:**
   - **Explanation:** MongoDB allows for the creation of custom indexes on fields, which significantly improves query performance. Indexes can be defined to speed up queries for specific data access patterns.

7. **Geospatial Capabilities:**
   - **Explanation:** MongoDB has built-in support for geospatial data and queries, making it an excellent choice for location-based applications. You can store and query geospatial data, such as coordinates and polygons, efficiently.

8. **Full-Text Search:**
   - **Explanation:** MongoDB includes a powerful text search feature that enables developers to perform full-text searches on textual data within documents. This feature is valuable for applications requiring advanced search functionality.

9. **Aggregation Framework:**
   - **Explanation:** MongoDB's Aggregation Framework allows developers to perform complex data transformations and computations within the database. It supports a wide range of aggregation stages, including grouping, sorting, filtering, and more.

10. **Security Features:**
    - **Explanation:** MongoDB offers robust security features, including authentication, authorization, encryption (both in transit and at rest), and auditing. It allows you to define access control at a granular level, ensuring data security.

11. **Schema Validation:**
    - **Explanation:** While MongoDB is schema-flexible, it provides optional schema validation rules that allow you to enforce data integrity and consistency by defining constraints on document structures.

12. **Community and Enterprise Editions:**
    - **Explanation:** MongoDB is available in both community and enterprise editions. The enterprise version includes additional features, support, and advanced tools for mission-critical applications.

13. **Rich Ecosystem:**
    - **Explanation:** MongoDB has a thriving ecosystem with official drivers and libraries for various programming languages, extensive documentation, and an active community. This makes it easier for developers to work with MongoDB in their preferred environment.

MongoDB's combination of flexibility, scalability, and rich features makes it a popular choice for a wide range of applications, from small startups to large enterprises, and across various industries.

#### Q3. Write a code to connect MongoDB to Python. Also, create a database and a collection in MongoDB.

To connect MongoDB to Python and create a database and a collection, you can use the PyMongo library, which is the official MongoDB driver for Python. If you haven't already installed PyMongo, you can do so using pip:

In [1]:
pip install pymongo

Note: you may need to restart the kernel to use updated packages.


In [2]:

from pymongo.mongo_client import MongoClient

uri = "mongodb+srv://shivamshrotria:shivam@cluster0.1eaisek.mongodb.net/?retryWrites=true&w=majority"

# Create a new client and connect to the server
client = MongoClient(uri)

# Send a ping to confirm a successful connection
try:
    client.admin.command('ping')
    print("Pinged your deployment. You successfully connected to MongoDB!")
except Exception as e:
    print(e)

Pinged your deployment. You successfully connected to MongoDB!


In [3]:
# Specify the name of the database you want to create or use
db_name = "mydatabase"

# Create a MongoDB database (if it doesn't already exist)
db = client[db_name]

# Specify the name of the collection you want to create or use
collection_name = "mycollection"

# Create a MongoDB collection (if it doesn't already exist)
collection = db[collection_name]

# Print a success message
print(f"Connected to MongoDB. Using database: {db_name}, collection: {collection_name}")

Connected to MongoDB. Using database: mydatabase, collection: mycollection


In [4]:
client

MongoClient(host=['ac-vojftt0-shard-00-01.1eaisek.mongodb.net:27017', 'ac-vojftt0-shard-00-00.1eaisek.mongodb.net:27017', 'ac-vojftt0-shard-00-02.1eaisek.mongodb.net:27017'], document_class=dict, tz_aware=False, connect=True, retrywrites=True, w='majority', authsource='admin', replicaset='atlas-rfkn4j-shard-0', tls=True)

In this code:

1. We import the `pymongo` library.
2. Define the MongoDB connection URL. Make sure to replace it with the actual URL of your MongoDB server. By default, MongoDB runs on port 27017 locally.
3. Create a `MongoClient` object to establish a connection to the MongoDB server.
4. Specify the name of the database you want to create or use (`db_name`).
5. Create a MongoDB database using the specified name (if it doesn't already exist).
6. Specify the name of the collection you want to create or use (`collection_name`).
7. Create a MongoDB collection using the specified name (if it doesn't already exist).
8. Print a success message indicating the connection and the selected database and collection.

You can then use the `collection` object to perform various database operations like inserting, querying, updating, and deleting documents in MongoDB.

#### Q4. Using the database and the collection created in question number 3, write a code to insert one record, and insert many records. Use the find() and find_one() methods to print the inserted record.

In [5]:
# Insert one record into the collection
record_one = {
    "name": "John",
    "age": 30,
    "city": "New York"
}

# Insert the record into the collection
inserted_record_id = collection.insert_one(record_one).inserted_id

In [6]:
for i in collection.find():
    print(i)

{'_id': ObjectId('64f86e106f6a38a8a78edf93'), 'name': 'Bob', 'age': 35, 'city': 'Los Angeles'}
{'_id': ObjectId('64f8b4366f6a38a8a78edf94'), 'name': 'Alice', 'age': 28, 'salary': 60000}
{'_id': ObjectId('64f8b4366f6a38a8a78edf95'), 'name': 'Bob', 'age': 35, 'salary': 75000}
{'_id': ObjectId('64f8b4366f6a38a8a78edf96'), 'name': 'Charlie', 'age': 22, 'salary': 55000}
{'_id': ObjectId('64f8b4366f6a38a8a78edf97'), 'name': 'David', 'age': 30, 'salary': 62000}
{'_id': ObjectId('64f8b5223f23feee57ce9692'), 'name': 'Alice', 'age': 25, 'city': 'San Francisco'}
{'_id': ObjectId('64f8b5223f23feee57ce9693'), 'name': 'Bob', 'age': 35, 'city': 'Los Angeles'}
{'_id': ObjectId('64f8b5223f23feee57ce9694'), 'name': 'Alice', 'age': 28, 'salary': 60000}
{'_id': ObjectId('64f8b5223f23feee57ce9695'), 'name': 'Bob', 'age': 35, 'salary': 75000}
{'_id': ObjectId('64f8b5223f23feee57ce9696'), 'name': 'Charlie', 'age': 22, 'salary': 55000}
{'_id': ObjectId('64f8b5223f23feee57ce9697'), 'name': 'David', 'age': 30, 

In [7]:
# # Print the inserted record
# print(f"Inserted record with ID: {inserted_record_id}")
# print("Inserted Record:")
# print(collection.find_one({"_id": inserted_record_id}))

# Insert multiple records into the collection
records_many = [
    {
        "name": "Alice",
        "age": 25,
        "city": "San Francisco"
    },
    {
        "name": "Bob",
        "age": 35,
        "city": "Los Angeles"
    }
]

# Insert the records into the collection
inserted_records_ids = collection.insert_many(records_many).inserted_ids

# Print the inserted records
print("\nInserted Records:")
for record_id in inserted_records_ids:
    print(collection.find_one({"name": 'Alice'}))



Inserted Records:
{'_id': ObjectId('64f8b4366f6a38a8a78edf94'), 'name': 'Alice', 'age': 28, 'salary': 60000}
{'_id': ObjectId('64f8b4366f6a38a8a78edf94'), 'name': 'Alice', 'age': 28, 'salary': 60000}


#### Q5. Explain how you can use the find() method to query the MongoDB database. Write a simple code to demonstrate this.

The `find()` method in MongoDB is used to query a collection and retrieve documents that match a specified query criteria. It returns a cursor, which you can iterate over to access the matching documents. Here's how you can use the `find()` method, along with a simple code example:
In this code:

1. We connect to the MongoDB server and access the specified database and collection.
2. We use the `find()` method on the collection to retrieve all documents in the collection. This query has no specific filter criteria, so it returns all documents.
3. The `find()` method returns a cursor that allows us to iterate over the matching documents.
4. We use a loop to iterate over the cursor and print each document, effectively displaying all documents in the collection.

For example, to find documents with a specific field value, you can do the following:

In [8]:
# Query the collection to find documents where the "city" field is "New York"
query = {"city": "New York"}
cursor = collection.find(query)

print("Documents with 'city' field set to 'New York':")
for document in cursor:
    print(document)

Documents with 'city' field set to 'New York':
{'_id': ObjectId('64f8b8372282acb314375334'), 'name': 'John', 'age': 30, 'city': 'New York'}


This code will find and print documents that have "New York" as the value of the "city" field. You can build more complex queries using various operators and criteria to filter and retrieve specific documents from your MongoDB collection.

#### Q6. Explain the sort() method. Give an example to demonstrate sorting in MongoDB.

The `sort()` method in MongoDB is used to sort the results of a query in ascending or descending order based on one or more fields in the documents. It allows you to specify the sorting criteria for the documents returned by a `find()` query. The `sort()` method takes one or more fields as parameters and specifies the sorting order for each field.

Here's the basic syntax of the `sort()` method:

```python
collection.find(query).sort(sort_criteria)
```

- `query`: This is an optional parameter that allows you to specify a query to filter the documents you want to sort. If not provided, all documents in the collection will be sorted.

- `sort_criteria`: This parameter specifies the sorting criteria. It can be a single field or a list of fields. For each field, you can specify the sorting order using `pymongo.ASCENDING` (for ascending order, which is the default) or `pymongo.DESCENDING` (for descending order).

Now, let's look at an example to demonstrate sorting in MongoDB:
Suppose you have a MongoDB collection named "employees" with documents like this:

In [9]:
doc = [
    {"name": "Alice", "age": 28, "salary": 60000},
    {"name": "Bob", "age": 35, "salary": 75000},
    {"name": "Charlie", "age": 22, "salary": 55000},
    {"name": "David", "age": 30, "salary": 62000}
]


collection.insert_many(doc)

If you want to retrieve the employees sorted by their age in ascending order, you can use the `sort()` method like this:

In [10]:
import pymongo
for i in collection.find().sort("age",pymongo.ASCENDING):
    print(i)

{'_id': ObjectId('64f8b4366f6a38a8a78edf96'), 'name': 'Charlie', 'age': 22, 'salary': 55000}
{'_id': ObjectId('64f8b5223f23feee57ce9696'), 'name': 'Charlie', 'age': 22, 'salary': 55000}
{'_id': ObjectId('64f8b8372282acb314375339'), 'name': 'Charlie', 'age': 22, 'salary': 55000}
{'_id': ObjectId('64f8b5223f23feee57ce9692'), 'name': 'Alice', 'age': 25, 'city': 'San Francisco'}
{'_id': ObjectId('64f8b8372282acb314375335'), 'name': 'Alice', 'age': 25, 'city': 'San Francisco'}
{'_id': ObjectId('64f8b4366f6a38a8a78edf94'), 'name': 'Alice', 'age': 28, 'salary': 60000}
{'_id': ObjectId('64f8b5223f23feee57ce9694'), 'name': 'Alice', 'age': 28, 'salary': 60000}
{'_id': ObjectId('64f8b8372282acb314375337'), 'name': 'Alice', 'age': 28, 'salary': 60000}
{'_id': ObjectId('64f8b4366f6a38a8a78edf97'), 'name': 'David', 'age': 30, 'salary': 62000}
{'_id': ObjectId('64f8b5223f23feee57ce9697'), 'name': 'David', 'age': 30, 'salary': 62000}
{'_id': ObjectId('64f8b8372282acb314375334'), 'name': 'John', 'age':

In this example, the `sort("age", pymongo.ASCENDING)` method sorts the documents in the "employees" collection based on the "age" field in ascending order. The result will be a list of employees sorted by their ages from youngest to oldest.

You can also sort by multiple fields by passing a list of field names and sorting orders to the `sort()` method, allowing you to create more complex sorting criteria.

#### Q7. Explain why delete_one(), delete_many(), and drop() is used.

In MongoDB, there are three methods for removing data from a collection: `delete_one()`, `delete_many()`, and `drop()`. Each of these methods serves a different purpose and should be used in specific situations:

1. **`delete_one(filter)`**:
   - **Purpose:** This method is used to delete a single document from a collection that matches the specified filter criteria.
   - **Use Case:** You would use `delete_one()` when you want to remove a specific document based on a particular condition, such as deleting a user's profile by their name.

   ```python
   collection.delete_one({"name": "Alice"})
   ```

In [11]:
collection.delete_one({"name": "Alice"})

In [12]:
for i in collection.find():
    print(i)

{'_id': ObjectId('64f86e106f6a38a8a78edf93'), 'name': 'Bob', 'age': 35, 'city': 'Los Angeles'}
{'_id': ObjectId('64f8b4366f6a38a8a78edf95'), 'name': 'Bob', 'age': 35, 'salary': 75000}
{'_id': ObjectId('64f8b4366f6a38a8a78edf96'), 'name': 'Charlie', 'age': 22, 'salary': 55000}
{'_id': ObjectId('64f8b4366f6a38a8a78edf97'), 'name': 'David', 'age': 30, 'salary': 62000}
{'_id': ObjectId('64f8b5223f23feee57ce9692'), 'name': 'Alice', 'age': 25, 'city': 'San Francisco'}
{'_id': ObjectId('64f8b5223f23feee57ce9693'), 'name': 'Bob', 'age': 35, 'city': 'Los Angeles'}
{'_id': ObjectId('64f8b5223f23feee57ce9694'), 'name': 'Alice', 'age': 28, 'salary': 60000}
{'_id': ObjectId('64f8b5223f23feee57ce9695'), 'name': 'Bob', 'age': 35, 'salary': 75000}
{'_id': ObjectId('64f8b5223f23feee57ce9696'), 'name': 'Charlie', 'age': 22, 'salary': 55000}
{'_id': ObjectId('64f8b5223f23feee57ce9697'), 'name': 'David', 'age': 30, 'salary': 62000}
{'_id': ObjectId('64f8b8372282acb314375334'), 'name': 'John', 'age': 30, '

2. **`delete_many(filter)`**:
   - **Purpose:** `delete_many()` is used to delete multiple documents from a collection that match the specified filter criteria.
   - **Use Case:** You would use `delete_many()` when you need to remove multiple documents that meet a specific condition, like deleting all documents where the "city field is "New York."

   ```python
   collection.delete_many({"city": "New York"})
   ```

In [13]:
collection.delete_many({"city": "New York"})

In [14]:
for i in collection.find():
    print(i)

{'_id': ObjectId('64f86e106f6a38a8a78edf93'), 'name': 'Bob', 'age': 35, 'city': 'Los Angeles'}
{'_id': ObjectId('64f8b4366f6a38a8a78edf95'), 'name': 'Bob', 'age': 35, 'salary': 75000}
{'_id': ObjectId('64f8b4366f6a38a8a78edf96'), 'name': 'Charlie', 'age': 22, 'salary': 55000}
{'_id': ObjectId('64f8b4366f6a38a8a78edf97'), 'name': 'David', 'age': 30, 'salary': 62000}
{'_id': ObjectId('64f8b5223f23feee57ce9692'), 'name': 'Alice', 'age': 25, 'city': 'San Francisco'}
{'_id': ObjectId('64f8b5223f23feee57ce9693'), 'name': 'Bob', 'age': 35, 'city': 'Los Angeles'}
{'_id': ObjectId('64f8b5223f23feee57ce9694'), 'name': 'Alice', 'age': 28, 'salary': 60000}
{'_id': ObjectId('64f8b5223f23feee57ce9695'), 'name': 'Bob', 'age': 35, 'salary': 75000}
{'_id': ObjectId('64f8b5223f23feee57ce9696'), 'name': 'Charlie', 'age': 22, 'salary': 55000}
{'_id': ObjectId('64f8b5223f23feee57ce9697'), 'name': 'David', 'age': 30, 'salary': 62000}
{'_id': ObjectId('64f8b8372282acb314375335'), 'name': 'Alice', 'age': 25, 

3. **`drop()`**:
   - **Purpose:** The `drop()` method is used to completely remove an entire collection, including all of its documents.
   - **Use Case:** You would use `drop()` when you want to delete an entire collection and its contents. This is a more drastic operation compared to `delete_one()` and `delete_many()` because it removes the collection itself, and you'll need to re-create the collection if you want to use it again.

   ```python
   collection.drop()
   ```

It's important to exercise caution when using the `drop()` method because it permanently deletes all data in the collection, and there is no way to undo this operation. Use it only when you are certain that you want to completely remove the collection.

In summary, `delete_one()` and `delete_many()` are used for fine-grained control when you want to remove specific documents from a collection based on filter criteria. `drop()`, on the other hand, is a more drastic operation that removes the entire collection and all its documents. The choice of which method to use depends on your specific data management needs and whether you want to remove individual documents or the entire collection.

## collection.drop()